src/Bicep.Core/Parser/Parser.cs

```diff
@@ -410,6 +449,33 @@ private IdentifierSyntax Identifier(DiagnosticBuilder.ErrorBuilderDelegate error
             return new IdentifierSyntax(identifier);
         }
 
+        private IdentifierSyntaxBase IdentifierWithRecovery(DiagnosticBuilder.ErrorBuilderDelegate errorFunc, params TokenType[] terminatingTypes)
+        {
+            var identifierOrSkipped = this.WithRecovery(
+                () => Identifier(errorFunc),
+                RecoveryFlags.None,
+                terminatingTypes);
+
+            switch (identifierOrSkipped)
+            {
+                case IdentifierSyntax identifier:
+                    return identifier;
+
+                case SkippedTriviaSyntax skipped when skipped.Elements.Any() == false:
+                    var position = this.reader.Peek().Span.Position;
+                    return new MalformedIdentifierSyntax(new[] {CreateMissingToken(position)}, skipped.Diagnostics);
+
+                case SkippedTriviaSyntax skipped:
+                    // WithRecovery only skips tokens, so the cast should be safe
+                    return new MalformedIdentifierSyntax(skipped.Elements.Cast<Token>(), skipped.Diagnostics);
+
+                default:
+                    throw new NotImplementedException($"Unexpected identifier syntax type '{identifierOrSkipped.GetType().Name}'");
+            }
+        }
+
+        private Token CreateMissingToken(int position) => new Token(TokenType.Missing, new TextSpan(position, 0), string.Empty, ImmutableArray<SyntaxTrivia>.Empty, ImmutableArray<SyntaxTrivia>.Empty);
```

Comment:
We have round-trip test for partially-parsed documents, right?

Reply:
Yup we do.